### PR TITLE
Fix, clean up, and unify documentation of elementary reflectors

### PR DESCRIPTION
**Pull request by vladimir-ch**
* Use zero-based indexing for elementary reflectors (almost) everywhere.
* Denote elementary reflectors as H(i), not H_i, not H[i].
* Remove some lint.

### CODE REVIEW

File: cgo/lapack.go [hunk 1]
```diff
@@ -410,12 +410,12 @@ func (impl Implementation) Dgelqf(m, n int, a []float64, lda int, tau, work []fl
 // The ith elementary reflector can be explicitly constructed by first extracting
 // the
 //  v[j] = 0           j < i
-//  v[j] = i           j == i
-//  v[j] = a[i*lda+j]  j > i
-// and computing h_i = I - tau[i] * v * v^T.
+//  v[j] = 1           j == i
+//  v[j] = a[j*lda+i]  j > i
```
Comment (vladimir-ch) [inline cgo/lapack.go]: I think that v is stored in the columns of a but please check this @btracey

File: lapack64/lapack64.go [hunk 2]
```diff
@@ -99,19 +99,19 @@ func Gels(trans blas.Transpose, a blas64.General, b blas64.General, work []float
...
 // this function will panic otherwise.
 //
 // The ith elementary reflector can be explicitly constructed by first extracting
 // the
 //  v[j] = 0           j < i
-//  v[j] = i           j == i
-//  v[j] = a[i*lda+j]  j > i
-// and computing h_i = I - tau[i] * v * v^T.
+//  v[j] = 1           j == i
+//  v[j] = a[j*lda+i]  j > i
```
Comment (vladimir-ch) [inline lapack64/lapack64.go]: Also here.

File: native/dgeqr2.go [hunk 3]
```diff
@@ -20,12 +20,12 @@ import "github.com/gonum/blas"
 // The ith elementary reflector can be explicitly constructed by first extracting
 // the
 //  v[j] = 0           j < i
-//  v[j] = i           j == i
-//  v[j] = a[i*lda+j]  j > i
-// and computing h_i = I - tau[i] * v * v^T.
+//  v[j] = 1           j == i
+//  v[j] = a[j*lda+i]  j > i
```
Comment (vladimir-ch) [inline native/dgeqr2.go]: And here.

File: native/dlatrd.go [hunk 4]
```diff
@@ -53,11 +53,11 @@ import (
 // H has the form
 //  I - tau * v * v^T
 // If uplo == blas.Upper,
-//  Q = H[n] * H[n-1] * ... * H[n-nb+1]
+//  Q = H(n) * H(n-1) * ... * H(n-nb+1)
```
Comment (vladimir-ch) [inline native/dlatrd.go]: This seems to be still 1-indexed but (at least at the moment) my head refuses to convert it to 0-based confidently, especially together with the description of v.

**Comment by kortschak** [comment]
What is the rationale behind the use of parens versus brackets? Do we have one?

**Comment by btracey** [comment]
I think it should be brackets. This implies the go style and not the fortran one. It's not strictly like the go one since they are columns, but it's a notional documentation anyway.

**Comment by vladimir-ch** [comment]
I have no strong preference, my intention is just to reduce inconsistencies in the documentation.
I would go with the Latex-like notation H_k but that becomes less clear when the index is an expression and using {} makes it look comlicated.
In the end I went with parens because brackets evoke slices and H are neither slices nor vectors. I kept the usage of brackets for things that have their direct equivalent in the code, for example the tau slice. I see parens here not as a Fortran notation, rather as a function notation: H is a function that returns elementary reflector.
But yes, it's a notational documentation with limited means of expression, a bit of a nitpicking, so I'm happy to change it to whatever you think will work best.

**Comment by kortschak** [comment]
If you say they're functions, then I'm happy with that, but I'll leave the decision to you and Brendan.

**Comment by vladimir-ch** [comment]
Looking at `H` as a matrix function of `i` is one possible way to justify using `()` over `[]`. Another justification could be reserving `[]` for indexing matrices, vectors, and slices, which is not the case 
of `H(i)`. Using `H_1` and `H_{n-i+1}` here would probably be most appropriate but I don't know how people not familiar with Latex perceive this notation.

To sum it up, what I'm trying to say: given that we have just ASCII and a limited subset of UTF8 symbols at hand to document a numerical library, having some basic set of conventions applied consistently would help having a clean and confidence-conveying documentation.

**Comment by btracey** [comment]
You're right, the underscore-brace system is probably better.

File: cgo/lapack.go (hunk 1 above)
Comment (btracey) [inline cgo/lapack.go]: Yes, columns.

File: lapack.go [hunk 5]
```diff
@@ -48,8 +48,8 @@ type Float64 interface {
 type Direct byte
 
 const (
-	Forward  Direct = 'F' // Reflectors are right-multiplied, H_1 * H_2 * ... * H_k
-	Backward Direct = 'B' // Reflectors are left-multiplied, H_k * ... * H_2 * H_1
+	Forward  Direct = 'F' // Reflectors are right-multiplied, H(0) * H(1) * ... * H(k).
```
Comment (btracey) [inline lapack.go]: I think needs to be k-1

File: lapack64/lapack64.go (hunk 2 above)
Comment (btracey) [inline lapack64/lapack64.go]: Yep

File: native/dlatrd.go (hunk 4 above)
Comment (btracey) [inline native/dlatrd.go]: It does seem to be one indexed. See line 78 of testlapack/dlatrd.

File: native/dlatrd.go [hunk 6]
```diff
@@ -53,11 +53,11 @@ import (
...
 // where v[:i-1] is stored in A[:i-1,i], v[i-1] = 1, and v[i:n] = 0.
 //
 // If uplo == blas.Lower,
-//  Q = H[1] * H[2] * ... H[nb]
+//  Q = H(1) * H(2) * ... H(nb)
```
Comment (btracey) [inline native/dlatrd.go]: This also seems one-indexed

File: cgo/lapack.go (hunk 1 above)
Comment (btracey) [inline cgo/lapack.go]: yep

**Comment by btracey** [comment]
LGTM, with the small corrections.

File: native/dlatrd.go (hunk 4 above)
Comment (vladimir-ch) [inline native/dlatrd.go]: The pointer to the test was very useful. Thanks. Done.

File: lapack.go (hunk 5 above)
Comment (vladimir-ch) [inline lapack.go]: Done, although here without context the final (or initial) subscript does not really matter.

File: native/dlatrd.go (hunk 6 above)
Comment (vladimir-ch) [inline native/dlatrd.go]: Done.

**Pull request merged**